[Kineto][NCCL][2/n] Add records NCCL meta to more collective functions

yoyoyocmu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111846
* __->__ #111843
* #111842

This diff records NCCL metadata for more commonly used collective functions.

NOTE: the coalesced NCCL are not covered: https://fburl.com/code/ihgqqvg8 and how to support them needs further discussion.

Differential Revision: [D50439232](https://our.internmc.facebook.com/intern/diff/D50439232/)

yoyoyocmu committed:
[Kineto][NCCL][2/n] Add records NCCL meta to more collective functions

This diff records NCCL metadata for more commonly used collective functions.

NOTE: the coalesced NCCL are not covered: https://fburl.com/code/ihgqqvg8 and how to support them needs further discussion.

Differential Revision: [D50439232](https://our.internmc.facebook.com/intern/diff/D50439232/)

[ghstack-poisoned]

aaronenyeshi commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2728,8 +2726,7 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::_reduce_scatter_base(
       "_reduce_scatter_base", // colName
-      tensor.numel() * // inSize
-          this->getSize(),
+      inputTensor.numel(), // inSize
```
wondering why we should change this, but not for gather on line 2470.

yoyoyocmu replied:
The change here can help us remove the unnecessary calculation. But for allgahter function on line 2470, there's no such straightforward way to get the result.

yoyoyocmu committed:
Update on "[Kineto][NCCL][2/n] Add records NCCL meta to more collective functions"


This diff records NCCL metadata for more commonly used collective functions.

NOTE: the coalesced NCCL are not covered: https://fburl.com/code/ihgqqvg8 and how to support them needs further discussion.

Differential Revision: [D50439232](https://our.internmc.facebook.com/intern/diff/D50439232/)

[ghstack-poisoned]
